### PR TITLE
[SYCL] Detect host built-in functions inside device code and emit a deferred diagnostic

### PR DESCRIPTION
**Pull request by Srividya2212**
As a follow up to the previous issue that @erichkeane fixed in llvm community,
this patch adds the deferred diagnostic to diagnose when a host builtin is used
in device mode.

- Added a deferred diagnostic that will be emitted whenever an host target built-in function is found inside device code.
- Added a test to verify that the deferred diagnostic is emitted correctly.

### CODE REVIEW

File: clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+int main(int argc, char **argv) {
+  //_mm_prefetch is an x86-64 specific builtin where the second integer parameter is required to be a constant
+  //between 0 and 7.
+  _mm_prefetch("test", 4); // no error thrown, since this is a valid invocation
+
+  _mm_prefetch("test", 8); // expected-error {{argument value 8 is outside the valid range [0, 7]}}
+
+  cl::sycl::kernel_single_task<class AName>([]() {
+    _mm_prefetch("test", 4); // expected-error {{Host specific builtin cannot be used in device functions}}
+    _mm_prefetch("test", 8); // expected-error {{argument value 8 is outside the valid range [0, 7]}} expected-error {{Host specific builtin cannot be used in device functions}}
```
Comment (Srividya2212) [inline clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp:25]: > Do this same line inside the kernel as well. It should have the same effect

I had to add both the expectations. When I had just `expected-error {{argument value 8 is outside the valid range [0, 7]}}`
It complained about the other missing expectation and vice versa.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -10878,4 +10878,6 @@ def err_ext_int_bad_size : Error<"%select{signed|unsigned}0 _ExtInt must "
                                  "have a bit size of at least %select{2|1}0">;
 def err_ext_int_max_size : Error<"%select{signed|unsigned}0 _ExtInt of bit "
                                  "sizes greater than %1 not supported">;
+def err_aux_target_builtin_in_device_code : Error<
+  "Host specific builtin cannot be used in device functions">;
```
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10882]: How about re-using existing diagnostic https://github.com/intel/llvm/blob/ae3fd5c48bb55176016a7c7c6fa2ac5684aaff55/clang/include/clang/Basic/DiagnosticSemaKinds.td#L9357 ?

File: clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -triple spir64-unknown-unknown-sycldevice -fsycl -fsycl-is-device -aux-triple x86_64-unknown-linux-gnu -verify -fsyntax-only  %s
+//
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp:2]:
```suggestion
```

File: clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+} // namespace sycl
+} // namespace cl
+
+int main(int argc, char **argv) {
+  //_mm_prefetch is an x86-64 specific builtin where the second integer parameter is required to be a constant
+  //between 0 and 7.
+  _mm_prefetch("test", 4); // no error thrown, since this is a valid invocation
+
+  _mm_prefetch("test", 8); // expected-error {{argument value 8 is outside the valid range [0, 7]}}
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp:21]: I don't think that you need to check this behavior in SYCL test.

File: clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp:25]: This is accurate.  The first one causes the host-specific builtin error, the second one causes the argument value issue first.

File: clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp (hunk 4 above)
Comment (erichkeane) [reply clang/test/SemaSYCL/deferred-diagnostics-aux-builtin.cpp:21]: I thought it was valuable here simply to illustrate that we're maintaining behavior.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
Comment (erichkeane) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10882]: Thats a good idea @Fznamznon.

File: clang/lib/Sema/SemaChecking.cpp [hunk 5]
```diff
@@ -1938,11 +1938,15 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
     if (Context.BuiltinInfo.isAuxBuiltinID(BuiltinID)) {
       assert(Context.getAuxTargetInfo() &&
              "Aux Target Builtin, but not an aux target?");
-
```
Comment (erichkeane) [inline clang/lib/Sema/SemaChecking.cpp:1941]: Don't make unrelated changes, just put htis line back in.

**Comment by Srividya2212** [comment]
@erichkeane  Who should merge this? I don't have merge permission.

**Comment by erichkeane** [comment]
> 
> 
> @erichkeane Who should merge this? I don't have merge permission.

@bader has to do it.

**Comment by bader** [comment]
> > @erichkeane Who should merge this? I don't have merge permission.
> 
> @bader has to do it.

It only looks like @bader has to do it, but we have a team of project maintainers, which includes 8(!) people, who can merge PRs.
